File: kryptografie/prg/Grid.php
<?php
class Grid
{
    private array $grid;
    private int $size;

    public function __construct()
    {
        $this->grid = [];
        $this->size = 0;
    }

    public function load(string $file): static
    {
        $this->grid = json_decode(
            file_get_contents(__DIR__ . '/' . $file),
            true
        );
        $this->size = count($this->grid);
        return $this;
    }

    public function rotate(): static
    {
        $rotated_grid = [];
        // transform the grid
        for ($y = 0; $y < $this->size; $y++) {
            for ($x = 0; $x < $this->size; $x++) {
                $rotated_grid[$y][$x] = $this->grid[$x][$y];
            }
        }
        // reverse the items in rows
        foreach ($rotated_grid as &$row) {
            $row = array_reverse($row);
        }
        $this->grid = $rotated_grid;
        return $this;
    }

    public function validate(): bool
    {
        $canvas = [];
        foreach (range(0, 3) as $state) {
            for ($y = 0; $y < $this->size; $y++) {
                for ($x = 0; $x < $this->size; $x++) {
                    if ($this->grid[$x][$y] == 0) {
                        if (!isset($canvas[$x][$y])) {
                            $canvas[$x][$y] = 1;
                        } else {
                            return false;
                        }
                    }
                }
            }
            $this->rotate();
        }
        return true;
    }

    public function encrypt(string $text): string
    {
        $canvas = [];
        $char_index = 0;
        $grid_shift = 0;
        $end = false;
        while (!$end) {
            foreach (range(0, 3) as $state) {
                for ($y = 0; $y < $this->size; $y++) {
                    for ($x = $grid_shift; $x < $grid_shift + $this->size; $x++) {
                        if (!$end && $this->grid[$y][$x - $grid_shift] == 0) {
                            $canvas[$y][$x] = $text[$char_index++];
                            if ($char_index == strlen($text)) {
                                $end = true;
                            }
                        } else if (!isset($canvas[$y][$x])) {
                            $canvas[$y][$x] = " ";
                        }
                    }
                }
                $this->rotate();
            }
            $grid_shift += $this->size;
        }
        $output = '';
        foreach ($canvas as $row) {
            $output .= implode('', $row) . "\n";
        }
        return $output;
    }

    public function __toString(): string
    {
        $output = "Grid:\n";
        foreach ($this->grid as $row) {
            $output .= " " . implode("", $row) . "\n";
        }
        return $output;
    }
}

File: kryptografie/prg/python/grid.py
class Grid:
    # kontruktor - načítá mřížku z csv souboru
    def __init__(self, file: str):
        with open("prg/python/" + file) as f:
            self.grid = [[int(num) for num in line.split(",")] for line in f]
        self.size = len(self.grid)

    # rotuje mřížku o 90° ve směru hodinových ručiček
    def rotate(self):
        self.grid = list(map(list, zip(*self.grid[::-1])))        

    # kontroluje, je-li mřížka validní
    def validate(self) -> bool:
        canvas = [([0] * self.size) for _ in range(self.size)]
        for state in range(4):
            for y in range(self.size):
                for x in range(self.size):
                    if self.grid[x][y] == 0:
                        if canvas[x][y] == 1:
                            return False
                        canvas[x][y] = 1                    
            self.rotate()
        return True

    def encrypt(self, text: str) -> str:
        # předpokládám optimální mřížku (max děr):
        multiplier = ((len(text)-1) // self.size**2) + 1
        canvas = [([" "] * self.size*multiplier) for _ in range(self.size)]
        text_index = 0
        grid_shift = 0
        end = False
        while not end:
            for state in range(4):
                for x in range(self.size):
                    for y in range(self.size):
                        if not end and self.grid[x][y] == 0:
                            canvas[x][y + grid_shift] = text[text_index]
                            text_index += 1
                            if text_index == len(text):
                                end = True                            
                self.rotate()                
            grid_shift += self.size            
        output = ""
        for x in range(len(canvas)):
            for y in range(len(canvas[0])):
                output += canvas[x][y]
            output += "\n"
        return output
        
    def print(self) -> str:
        output = "Grid:\n"
        for x in range(self.size):
            for y in range(self.size):
                output += str(self.grid[x][y])
            output += "\n"
        print(output)
